source/Octopus.Tentacle/Services/Scripts/ScriptServiceV2.cs

```diff
@@ -0,0 +1,163 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Threading;
+using Octopus.Diagnostics;
+using Octopus.Tentacle.Contracts;
+using Octopus.Tentacle.Scripts;
+
+namespace Octopus.Tentacle.Services.Scripts
+{
+    [Service]
+    public class ScriptServiceV2 : IScriptServiceV2
+    {
+        readonly IShell shell;
+        readonly IScriptWorkspaceFactory workspaceFactory;
+        readonly IScriptStateStoreFactory scriptStateStoreFactory;
+        readonly ISystemLog log;
+        readonly ConcurrentDictionary<ScriptTicket, RunningScriptWrapper> runningScripts = new();
+
+        public ScriptServiceV2(
+            IShell shell,
+            IScriptWorkspaceFactory workspaceFactory,
+            IScriptStateStoreFactory scriptStateStoreFactory,
+            ISystemLog log)
+        {
+            this.shell = shell;
+            this.workspaceFactory = workspaceFactory;
+            this.scriptStateStoreFactory = scriptStateStoreFactory;
+            this.log = log;
+        }
+
+        public ScriptStatusResponseV2 StartScript(StartScriptCommandV2 command)
+        {
+            var runningScript = runningScripts.GetOrAdd(
+                command.ScriptTicket,
+                s =>
+                {
+                    var workspace = workspaceFactory.GetWorkspace(command.ScriptTicket);
+                    var scriptState = scriptStateStoreFactory.Get(s, workspace);
+                    return new RunningScriptWrapper(scriptState);
+                });
+
+            using (runningScript.ScriptStateStore.GetExclusiveLock())
```

Comment:
Should this lock be be in the `ScriptStateStore`? Is the lock tied to the store?

I wonder if instead the lock should live in `runningScripts` inside of `RunningScriptWrapper`.  It could be called `StartScriptMutex`.

Reply:
Yeh makes sense 👍